### ISSUE TITLE
forge: Deploying and testing contracts compiled by older Solidity versions

### ISSUE THREAD

**Issue by Jeiwan** [gakonst/foundry#326]
When compiling a test contract that imports a contract that uses an older Solidity version, it fails with:
> Error:
   0:
      ParserError: Source file requires different compiler version (current compiler is 0.8.10+commit.fc410830.Darwin.appleclang) - note that nightly builds are considered to be strictly less than the released version
       --> /Users/jeiwan/.tmp/foundrysolversions/lib/v2-core/contracts/UniswapV2Pair.sol:1:1:
        |
      1 | pragma solidity =0.5.16;
        |
        
Steps to reproduce:
1. `mkdir foundrysolversions && cd $_`
1. `forge init`
1. `forge install Uniswap/v2-core`
1. Edit `src/test/Contract.t.sol` and add `import "v2-core/UniswapV2Pair.sol";`
1. `forge test`

Is there a workaround? Is it possible to compile contracts with different compilers and then reuse binary code?

**Comment by gakonst** [comment]
I don't think this is possible in general? You should not be able to import & use a file that is `pragma solidity =0.5.16` in a non -0.5.16 file. 

2 workarounds:
1. You could write an interface of the functions you want to call that is compatible with 0.8 or whatever you're using
2. You could change your contracts to use 0.5.16

I'd recommend the former.

**Comment by mattsse** [comment]
`pragma solidity =0.5.16;` pins to `0.5.16` and `0.5.16` only, meaning it won't compile with any other version.

this leaves two possible solutions:
* use the exact version yourself (or a range that includes it like >=0.5.16)
* upgrade the contract's pragma to a newer version 

There is an unrecommended third possible workaround:
technically, it's not necessarily required to have a version pragma in a solidity file, but it's highly! recommended and the defacto standard. depending on the build tool you don't need a version pragma in your imports.

**Comment by gakonst** [comment]
Yep, matches my understanding above. Closing then as it's not a Foundry issue and isn't something support-able in any framework afaict.

**Comment by Jeiwan** [comment]
This is doable in Hardhat, although with a hack. I think it's also possible to deploy binary code compiled by older Solidity compiler as a contract using inline assembly or some other magic. Probably this can be implemented as another cheat code (the VM contract).

This feature can be needed when testint a contract that interacts with third-party contracts deployed on mainnet. For example, a contract that uses flash loans of Uniswap. Upgrading outdated contracts can cause bugs or result in different behavior. Also, since Forge dependencies are git submodule, one would have to fork original repo and add it as a dependency instead of the original one, to be able to save changes.

Mocking is a solution but mocking is generally considered a bad practice because mocks are not exact copies and can differ in behavior. For simple cases like flash loans mocks are ok, but for more complex cases they can be problematic: imagine you used a slightly incorrect mock and then deployed to mainnet where the contract you depend on works differently. At this point though mocks seems like the simplest solution.

**Comment by gakonst** [comment]
What is the way it's doable in hardhat? Maybe we can streamline that hack?

**Comment by Jeiwan** [comment]
First step is to compile a dependency contract separately from other contracts. In Hardhat, I use an empty contract like:
```solidity
pragma solidity =0.5.16;

import "@uniswap/v2-core/contracts/UniswapV2Factory.sol";

contract DependencyUniswapV2 {}
```
Then, in tests, it can be easily used via:
```js
const Factory = await ethers.getContractFactory("UniswapV2Factory");
```
`getContractFactory` is a feature of Hardhat, it loads contracts from artifacts by contract name and it can instantiate a contract from bytecode and ABI:
https://github.com/nomiclabs/hardhat/blob/767f68dee84d6a47cd40153a8ab8552f41a95d5e/packages/hardhat-ethers/src/internal/helpers.ts#L60
I wish we could have something like that in Forge because writing tests in Solidity is so much better than in JS. It also feels like making Solidity compiler import contracts from bytecode is not a big problem, but it might be easier to hack it in Forge first.

**Comment by jparklev** [comment]
Here's what i'm doing for this atm lmao 

<img width="747" alt="Screen Shot 2021-12-31 at 2 40 11 PM" src="https://user-images.githubusercontent.com/24902242/147837346-11d88719-ec94-4f19-8a25-644b992af4ee.png">

**Comment by gakonst** [comment]
Honestly that's not a bad workaround, maybe something we should add on the stdlib?

2 qs:
1. What does _compileSpace() do?
2. How do you get its interface for further usage in the tests?

**Comment by jparklev** [comment]
> What does _compileSpace() do?

I'm using a monorepo, so its essentially making sure the other pkg has been compiled first (does something like `cd ../space-pkg; forge build`)

> How do you get its interface for further usage in the tests?

Just defining a new interface for it [maker-style](https://github.com/makerdao/dss/blob/c666ab1fdac4cb3dd8a8b4223f951a9773a64c55/src/pot.sol#L45) in this case, nothing fancy

> maybe something we should add on the stdlib

Yea! Guess it depends if there are any gotchas with appending calldata to bin bytecodes compiled with different versions? Though I don't know of anything like that, I would defer to others who are more familiar

**Comment by Jeiwan** [comment]
I've ended up using the same solution. Learned about it from this blog post:
https://www.libevm.com/2021/12/18/yulp-smart-contract-dev/

**Comment by brockelmore** [comment]
May i suggest to reduce external dependencies like `jq`, that we make it a cheatcode for loading deplyoment (and runtime) bytecode from a compiled contract? This would also make the `etch` cheatcode more useful i imagine

**Comment by gakonst** [comment]
Supportive @brockelmore

**Comment by ncitron** [comment]
If nobody else has, I'd like to give this one a shot. I've been wanting this feature for a while now in dapptools/foundry.

**Comment by onbjerg** [comment]
Is this solved by #440?

**Comment by Jeiwan** [comment]
@onbjerg Yes! Thanks.
